### PR TITLE
Refactoring for safer olm and megolm session usage

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Rework following investigations on strange rolling UISIs errors.
Logs contained a lot of BAD_MAC errors on both olm and megolm.

Olm bad macs can pretty much be anything, like using wrong/desync session or concurrency, or just false positive logs when iterating other known sessions (though proliferation of session might be suspicious)

Megolm could be more probably concurrent decryption, as well as desync with cache/store, as the current code was not very safe.

- The code has been refactored to create spefic classes to handle cache of sessions, and move this responsability out of the realm crypto store. Access are guarded by synchronized directive.
- Concurrent decryption attempt are guarded with mutex (per olm/megolm session)
- Access to olm account has been protected against concurrent access.
- Refactor the addInboundroupSession logic (there was some crash report of NPE there) and code could cause cache beeing out of sync
- Better logging
- *EDIT* Cleaner unwedging with better logs. Unwedging done at the end of the sync now,  some synchronisation added. 

Added a new E2E sanity test that the following is working:
- Proper key sharing to a set of user in e2e room
- Proper key backup restore from a new session
- Proper key forwarding
- Accept and replace existing session with a better forwarded one
 
 *EDIT* Resurrected unwedging test

### CODE REVIEW

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt
Comment (bmarty) [inline matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt:78]: 👋

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt
Comment (bmarty) [inline matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt:640]: Nit: Could use `assertNotNull`

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt
Comment (bmarty) [inline matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt:642]: I though there was an error, because the `message` should be the first parameter, but there are some sort of method aliases indeed:

image

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/PreShareKeysTest.kt
Comment (bmarty) [inline matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/PreShareKeysTest.kt:93]: If it was the source of the failure, that's a good catch, even if just running the test should show the error rather clearly :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt:153]: To avoid future mistake I would move the fun `doWithOlmAccount` to the store and remove getOlmAccount(), WDYT?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/OlmSessionWrapper.kt
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/OlmSessionWrapper.kt:31]: Could be a `val`

**Review by bmarty — COMMENTED** [review]
That's amazing work. I am not sure I understand all the code related to the crypto part. I am glad that we have some unit test. And you added lots of unit test which is also very nice. Thanks!
I would be more confident if we can have a review from someone in the crypto team. WDYT? @poljar maybe?

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/PreShareKeysTest.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/PreShareKeysTest.kt:93]: I didn't check others, but there are prob more

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt:640]: Done

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/internal/crypto/E2eeSanityTests.kt:642]: :/

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt:153]: yeah good idea

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/OlmSessionWrapper.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/model/OlmSessionWrapper.kt:31]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/algorithms/megolm/MXMegolmEncryption.kt
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/algorithms/megolm/MXMegolmEncryption.kt:474]: Is there a point to this `false.also`? Why not put the log line above the return and just `return false`? The same `log -> return` pattern is happening a couple lines bellow but `also {}` wasn't used there.

Perhaps we should either use `also {}` in such scenarios or not, but not mix/match.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt:869]: Another one, let's not keep cryptically commented code in the tree.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt
Comment (poljar) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt:283]: Let's remove this or document why it's still here.

**Review by poljar — APPROVED** [review]
Left some small nits, logic wise nothing seems to be wrong.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmSessionStore.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmSessionStore.kt:65]: mmm, initialy it was called also in another method (that was already synchronized), so I created this. But looks like there is no point now

Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmSessionStore.kt:65]: Done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmSessionStore.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmSessionStore.kt:82]: fixed (and some more in crypto store interface)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmSessionStore.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmSessionStore.kt:87]: Fixed

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/algorithms/megolm/MXMegolmEncryption.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/algorithms/megolm/MXMegolmEncryption.kt:474]: yes fixed

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt:869]: done

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt:283]: done

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/common/TestConstants.kt
Comment (michaelkaye) [inline matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/common/TestConstants.kt:26]: I found myself doing this in response to slow integration tests in #5459 - do you have any more information on why 60s is better?

**Review by michaelkaye — COMMENTED** [review]
I've recently commented out some flaky integration tests that are in this area: https://github.com/vector-im/element-android/issues/5449 - they seemed to be some sort of issue between two things opening a realm db at the same time. Do you think these concurrent access changes you  mention might have helped there - is it worth un-ignoring those tests?

Also, it might be good to run the nightly integration tests against a PR like this that introduces more integration tests - otherwise we won't compile and run them until the night after merge.

You can do that in Actions -> "Nightly build" (in left sidebar) -> then in the blue bar, select this branch to run the workflow against.

It doesn't run them automatically any more as it takes a good long time and we don't want to slow down the PR merge time too much.

File: matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/common/TestConstants.kt
Comment (BillCarsonFr) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/common/TestConstants.kt:26]: When doing test with several sessions everything is getting a lot slower (like E2E sanity test), each session is slower to sync, etc...
Everytime we test something with await/waitWithLatch it's using that as default timeout, and test are then failing when waiting from something coming back from the sync.
It also happens when running all tests, I wonder if we don't clean properly between test?
Previously I used the `ANDROIDX_TEST_ORCHESTRATOR` option to ensure that all test start with a clear state, but looks like it's not working anymore.

Comment (michaelkaye) [reply matrix-sdk-android/src/androidTest/java/org/matrix/android/sdk/common/TestConstants.kt:26]: We definitely don't clean up the server between tests (and if you just run a demo server locally, then that won't be cleaned up unless you explicitly do so)

**Review by bmarty — APPROVED** [review]
Let's merge this PR, 1.4.4 has been released.

**Pull request merged**
